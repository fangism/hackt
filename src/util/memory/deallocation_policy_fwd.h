/**
	\file "util/memory/deallocation_policy_fwd.h"
	Forward declarations of names of deallocation policies.  
	Their definitions appear in "util/memory/deallocation_policy.h".
	$Id: deallocation_policy_fwd.h,v 1.4 2007/03/12 07:38:15 fang Exp $
 */

#ifndef	__UTIL_MEMORY_DEALLOCATION_POLICY_FWD_H__
#define	__UTIL_MEMORY_DEALLOCATION_POLICY_FWD_H__

#include "util/sfinae_types.h"
// #include <cassert>		// for assert debugging

namespace util {
namespace memory {
//=============================================================================
/**
	Deallocate something allocated by new, single object allocation.
	This is the default assumed by all pointer classes.  
	\sa delete_array_tag.  
 */
struct delete_tag {
	template <class T>
	inline
	void
	operator () (T* t) const {
		delete t;
	}
};      // end struct delete_tag

//-----------------------------------------------------------------------------
/**
	Uses SFINAE to determine whether or not class has operator delete[].
 */
template <class T>
class has_class_operator_delete_array : public sfinae_types {
	template <void (*PF)(void*)>
	struct wrap_type { };	// what about throw specification?

	template <class U>
	static one __test(wrap_type<&U::operator delete[]>*);

	template <class U>
	static two __test(...);

public:
	enum { value = (sizeof(__test<T>(0)) == 1) } ;
};	// end class has_class_operator_delete_array

//- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
template <class, bool>
class array_deleter;

template <class T>
struct array_deleter<T, true> {
	inline
	void
	operator () (T* t) {
		T::operator delete [] (t);
	}
};	// end struct array_deleter<T, true>

template <class T>
struct array_deleter<T, false> {
	inline
	void
	operator () (T* t) {
		delete [] t;
	}
};	// end struct array_deleter<T, false>

template <class T>
inline
void
operator_delete_array(T* t) {
	array_deleter<T, has_class_operator_delete_array<T>::value>()(t);
}

//- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
/**
	Deallocate something allocated by new [], object array allocations.
	\sa delete_tag.
 */
struct delete_array_tag {
	template <class T>
	inline
	void
	operator () (T* t) const {
#if 0
		delete [] t;
#else
		operator_delete_array(t);
#endif
	}
};      // end struct delete_array_tag

//=============================================================================
struct free_tag;
typedef	free_tag	malloc_tag;

/**
	Standard FILE I/O streams.  
 */
struct fclose_tag;
typedef	fclose_tag	FILE_tag;

/**
	Process I/O file stream.  
 */
struct pclose_tag;
typedef	pclose_tag	process_tag;

struct iostream_tag;

template <class T, void (*f)(T*)>
struct custom_ptr_fun_tag;

template <class T, void (f)(T*)>
struct custom_fun_ref_tag;

template <class F>
struct custom_functor_tag;

//=============================================================================
}	// end namespace memory
}	// end namespace util

#endif	// __UTIL_MEMORY_DEALLOCATION_POLICY_FWD_H__

