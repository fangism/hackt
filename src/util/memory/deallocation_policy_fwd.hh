/**
	\file "util/memory/deallocation_policy_fwd.hh"
	Forward declarations of names of deallocation policies.  
	Their definitions appear in "util/memory/deallocation_policy.hh".
	$Id: deallocation_policy_fwd.hh,v 1.6 2011/02/04 02:23:42 fang Exp $
 */

#ifndef	__UTIL_MEMORY_DEALLOCATION_POLICY_FWD_H__
#define	__UTIL_MEMORY_DEALLOCATION_POLICY_FWD_H__

/**
	Should be 1, but gcc-3.3.x ICEs during attempted substitution.
	So we just disable it, there's no harm.  
 */
#define	UTIL_USE_SFINAE_OPERATOR_DELETE_ARRAY		0

#if UTIL_USE_SFINAE_OPERATOR_DELETE_ARRAY
#include "util/sfinae_types.hh"
#endif

namespace util {
namespace memory {
//=============================================================================
/**
	Deallocate something allocated by new, single object allocation.
	This is the default assumed by all pointer classes.  
	\sa delete_array_tag.  
 */
struct delete_tag {
	template <class T>
	inline
	void
	operator () (T* t) const {
		delete t;
	}
};      // end struct delete_tag

//-----------------------------------------------------------------------------
#if UTIL_USE_SFINAE_OPERATOR_DELETE_ARRAY
/**
	Uses SFINAE to determine whether or not class has operator delete[].
 */
template <class T>
class has_class_operator_delete_array : public sfinae_types {
	template <void (*PF)(void*)>
	struct wrap_type { };	// what about throw specification?

	template <class U>
	static one __test(wrap_type<&U::operator delete[]>*);

	template <class U>
	static two __test(...);

public:
	enum { value = (sizeof(__test<T>(0)) == 1) } ;
};	// end class has_class_operator_delete_array

//- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
template <class, bool>
class array_deleter;

template <class T>
struct array_deleter<T, true> {
	inline
	void
	operator () (T* t) {
		T::operator delete [] (t);
	}
};	// end struct array_deleter<T, true>

template <class T>
struct array_deleter<T, false> {
	inline
	void
	operator () (T* t) {
		delete [] t;
	}
};	// end struct array_deleter<T, false>

template <class T>
inline
void
operator_delete_array(T* t) {
	array_deleter<T, has_class_operator_delete_array<T>::value>()(t);
}
#endif	// UTIL_USE_SFINAE_OPERATOR_DELETE_ARRAY

//- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
/**
	Deallocate something allocated by new [], object array allocations.
	\sa delete_tag.
 */
struct delete_array_tag {
	template <class T>
	inline
	void
	operator () (T* t) const {
#if UTIL_USE_SFINAE_OPERATOR_DELETE_ARRAY
		operator_delete_array(t);
#else
		delete [] t;
#endif
	}
};      // end struct delete_array_tag

//=============================================================================
struct free_tag;
typedef	free_tag	malloc_tag;

/**
	Standard FILE I/O streams.  
 */
struct fclose_tag;
typedef	fclose_tag	FILE_tag;

/**
	Process I/O file stream.  
 */
struct pclose_tag;
typedef	pclose_tag	process_tag;

struct istream_tag;
struct ostream_tag;
// struct iostream_tag;

template <class T, void (*f)(T*)>
struct custom_ptr_fun_tag;

template <class T, void (f)(T*)>
struct custom_fun_ref_tag;

template <class F>
struct custom_functor_tag;

//=============================================================================
}	// end namespace memory
}	// end namespace util

#endif	// __UTIL_MEMORY_DEALLOCATION_POLICY_FWD_H__

