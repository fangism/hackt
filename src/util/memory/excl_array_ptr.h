/**
	\file "util/memory/excl_array_ptr.h"
	Exclusive-ownership pointers that are array allocated (new []).  
	Provides pointer classes:
		excl_array_ptr
		sticky_array_ptr
		some_array_ptr
		unique_array_ptr
	TODO: define the [] operator conditionally for array pointers
		in the base "excl_ptr.h"
	$Id: excl_array_ptr.h,v 1.1.4.1 2006/01/18 06:25:13 fang Exp $
 */

#ifndef	__UTIL_MEMORY_EXCL_ARRAY_PTR_H__
#define	__UTIL_MEMORY_EXCL_ARRAY_PTR_H__


#include "util/memory/excl_ptr.h"
#include "util/memory/deallocation_policy_fwd.h"

namespace util {
namespace memory {

template <class T>
struct excl_array_ptr {
	typedef	excl_ptr<T, delete_array_tag>		type;
};

}	// end namespace memory
}	// end namespace util

#endif	// __UTIL_MEMORY_EXCL_ARRAY_PTR_H__

