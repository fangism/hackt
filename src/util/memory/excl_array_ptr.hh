/**
	\file "util/memory/excl_array_ptr.hh"
	Exclusive-ownership pointers that are array allocated (new []).  
	Provides pointer classes:
		excl_array_ptr
		sticky_array_ptr
		some_array_ptr
		unique_array_ptr
	TODO: define the [] operator conditionally for array pointers
		in the base "excl_ptr.hh"
	$Id: excl_array_ptr.hh,v 1.2 2006/01/22 06:53:42 fang Exp $
 */

#ifndef	__UTIL_MEMORY_EXCL_ARRAY_PTR_H__
#define	__UTIL_MEMORY_EXCL_ARRAY_PTR_H__


#include "util/memory/excl_ptr.hh"
#include "util/memory/deallocation_policy_fwd.hh"

namespace util {
namespace memory {

template <class T>
struct excl_array_ptr {
	typedef	excl_ptr<T, delete_array_tag>		type;
};

}	// end namespace memory
}	// end namespace util

#endif	// __UTIL_MEMORY_EXCL_ARRAY_PTR_H__

