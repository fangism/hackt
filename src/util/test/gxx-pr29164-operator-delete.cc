/**
	\file "util/test/gxx-pr29164-operator-delete.cc"
	Bug test case, jacked up to test a workaround.  
	Suspicion: may not be a bug, call to class operator delete []
	is elided when pointer is NULL, not miscompiled.  
	Incidentally, several compilers are reported to behave similarly.  
	This test case will detect if behavior ever changes.  
	$Id: gxx-pr29164-operator-delete.cc,v 1.1 2007/03/12 07:38:17 fang Exp $
 */

#ifdef	NDEBUG
#undef	NDEBUG
#endif	// NDEBUG

#define	USE_EXCL_ARRAY_PTR		1

#include <iostream>
#if USE_EXCL_ARRAY_PTR
#include "util/memory/excl_array_ptr.h"
#else
#include "util/memory/deallocation_policy_fwd.h"
#endif
#include "one_array_only.h"

using std::cout;

//=============================================================================
#if USE_EXCL_ARRAY_PTR
using util::memory::excl_array_ptr;
#endif

int
main(int argc, char* argv[]) {
{
	one_array_only* p = new one_array_only[12];
//	delete [] p;		// triggers bug
//	one_array_only::operator delete [](p);	// works around bug
	util::memory::operator_delete_array(p);	// also works around bug
}
#if USE_EXCL_ARRAY_PTR
	const excl_array_ptr<one_array_only>::type
		array_ptr(new one_array_only[64]);
	// here not expected to call because excl_ptr suppresses
#endif
	return 0;
}


