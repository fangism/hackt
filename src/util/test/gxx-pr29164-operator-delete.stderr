class new[] operator
class delete[] operator
class new[] operator
